### ISSUE TITLE
[Feature Request] Stop workers for a specified amount of time after n 0, 0, 0 upserts

### ISSUE THREAD

**Issue by marcmicalizzi** [PokemonGoMap/PokemonGo-Map#506]
It would be valuable, to avoid full ip bans, to have a feature that would stop workers for a specified amount of time (say 15 minutes, but configurable) after a certain number of 0 pokemon, 0 pokestop, 0 gym upserts (indicating that the worker account has a softban).

It would also make sense to ramp up the stop time from 15 minutes to 30 minutes, and so forth, if the account still appears to be softbanned after the initial stoppage time.

**Comment by R0adRunn3r** [comment]
Yeah I opened one on this too but @BaIthamel closed it... I think that preventing IP Bans MUST be a priority! We can have a nice map with nice features but if the IP that is running on gets perma banned every 2 days is pointless! 

I ran the map in a small city for about 200 ppl and i use less than 10 accounts... my server (VPS) ip got banned... and the sign BEFORE the perma ban was that the workers found 0/0/0 for some searches 

IMO this is an HIGH priority thing to add, obviously as a choice since some people ran this on rural areas with genuine 0/0/0 findings

**Comment by Plop** [comment]
Yes please. This would be amazing.

**Comment by AndrewYoungs** [comment]
Pretty sure my VPS got soft banned today, perhaps proxy integration could be a thing?

**Comment by marcmicalizzi** [comment]
@AndrewYoungs proxy integration is completely unrelated to this ticket. that would be a separate feature altogether.

**Comment by R0adRunn3r** [comment]
@AndrewYoungs If I don't remember wrong proxy was added a couples of days ago

**Comment by AndrewYoungs** [comment]
@marcmicalizzi It's not completely unrelated if the IP block is causing 0/0/0 it can be allowed on PTC but the issue could be on Niantic's side. 

Also @marcmicalizzi since reading isn't a strong suit of yours "I think that preventing IP Bans MUST be a priority! We can have a nice map with nice features but if the IP that is running on gets perma banned every 2 days is pointless!" Seems pretty relevant. 

Also @R0adRunn3r I'll dig through commit notes and see if it was. I'd be interested in seeing if this would bypass the 0/0/0 returns.

**Comment by marcmicalizzi** [comment]
@AndrewYoungs unnecessary to resort to ad homenim insults. I was simply saying that this ticket is about adding a feature to stop workers when they encounter 0/0/0, not adding proxy support, so in the event that it is not yet added as a feature as mentioned by @R0adRunn3r, it would be better suited as a separate ticket, not a comment on a different feature request.

**Comment by reyqn** [comment]
Do we know what is the limit not to get soft ip banned? I don't think waiting for the end of the ban is really efficient. Maybe we should just try to avoid it.
From my experience, 2 requests / second is okay for the same ip (20 accounts, 10s delay)
20 requests / second got me soft ip banned (200 accounts, 10s delay).
Does someone have any other numbers?

**Comment by SOLiNARY** [comment]
@reyqn I have 17 workers/accounts on same IP with scan delay 10. not banned.

**Comment by Bulgran** [comment]
@reyqn I have 12 accounts with sd 10 on 1 IP through VPN and 6 accounts with sd 10 on another IP without issues.

**Comment by marcmicalizzi** [comment]
@reyqn The problem is it's really inconsistent. I have 3 servers on
different ips, one with 60 accounts, one with 70, one with 80, all running
10 separate -st 10 -sd 10 maps. 2 days ago soft bans would roll in after 2
hours, yesterday I got one softban in the morning, but none again for the
rest of the day/night. Some people get softbanned with a single worker. The
point is if you're not sitting at the computer and the softban comes in,
and you don't stop the process, the softban can turn into a 24 hour softban
instead of just 15 minutes if you stopped the worker. Since most people
would be using the map on the go, not viewing the logs, having an option to
automatically stop and start workers around softbans would be invaluable.

**Comment by R0adRunn3r** [comment]
You are lucky with soft bans... My server IP is perma banned... after 3 days still can't perform a scan with that IP

**Comment by reyqn** [comment]
@marcmicalizzi well that doesn't sound like good news :/ guess we shouldn't try to scan big zones anymore until someone figures out what exactly causes those bans...

**Comment by marcmicalizzi** [comment]
@reyqn doesn't even have to be very big though for some people to get banned. right now I don't even run the map if I can't kill it remotely easily, and if I'm not actively playing I don't keep it running either, because the chance of a softban starting to creep in and then me not killing the processes in time is too high.

**Comment by reyqn** [comment]
I was wondering if maybe we get banned because all of the accounts are querying the api at the same moment.
For example, when I use 20 accounts with a 10s delay, after 10s, all of my accounts send requests approximately at the same time. Maybe we should delay threads so that all accounts send a request every `scan_delay/number_of_accounts` seconds (in my case, I would then send a request regularly every 0.5s.
This would prevent big chunks of requests at the same instant, which could be the cause of the ban.

**Comment by marcmicalizzi** [comment]
@reyqn that doesn't match up with my going a whole day without a softban, or even how prior to that it would take at least 2 hours for the softbans to start rolling in though. What's interesting is that the ip doesn't get banned right off the bat. What I observed happening every time is one account gets banned, then a minute later or so the next, then the next and so on. The only time I ever actually experienced any ip ban (being I couldn't use the app on wifi but could on cellular), was when I left it running without watching for bans and all the accounts were running with softbans for a period of time (being they were all returning 0/0/0). Any other time, if I stopped scanning after 30-40% of the accounts had softbans, 15 minutes later they all worked again.
